File: internal/commands/analyze.go
package commands

import (
	"fmt"

	"github.com/spf13/cobra"
)

// AnalyzeCmd creates the analyze command
func AnalyzeCmd() *cobra.Command {
	analyzeCmd := &cobra.Command{
		Use:   "analyze",
		Short: "Analyze Kubernetes clusters and resources",
		Long: `Analyze Kubernetes clusters and resources for optimization opportunities.
		
Examples:
  upid analyze cluster                    # Analyze entire cluster
  upid analyze pod my-pod --namespace default  # Analyze specific pod
  upid analyze idle --confidence 0.85    # Find idle workloads
  upid analyze resources --time-range 24h # Analyze resource usage`,
		RunE: func(cmd *cobra.Command, args []string) error {
			return analyzeCluster(cmd, args)
		},
	}

	// Add subcommands
	analyzeCmd.AddCommand(analyzeClusterCmd())
	analyzeCmd.AddCommand(analyzePodCmd())
	analyzeCmd.AddCommand(analyzeIdleCmd())
	analyzeCmd.AddCommand(analyzeResourcesCmd())
	analyzeCmd.AddCommand(analyzeCostCmd())
	analyzeCmd.AddCommand(analyzePerformanceCmd())

	return analyzeCmd
}

// analyzeClusterCmd creates the cluster analysis command
func analyzeClusterCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "cluster [cluster-name]",
		Short: "Analyze Kubernetes cluster",
		Long:  "Perform comprehensive analysis of a Kubernetes cluster",
		RunE: func(cmd *cobra.Command, args []string) error {
			return analyzeCluster(cmd, args)
		},
	}

	// Add flags
	cmd.Flags().StringP("namespace", "n", "", "namespace to analyze")
	cmd.Flags().StringP("time-range", "t", "24h", "time range for analysis")
	cmd.Flags().BoolP("detailed", "d", false, "detailed analysis")
	cmd.Flags().BoolP("include-costs", "c", false, "include cost analysis")

	return cmd
}

// analyzePodCmd creates the pod analysis command
func analyzePodCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "pod [pod-name]",
		Short: "Analyze specific pod",
		Long:  "Analyze resource usage and performance of a specific pod",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return analyzePod(cmd, args)
		},
	}

	// Add flags
	cmd.Flags().StringP("namespace", "n", "default", "namespace of the pod")
	cmd.Flags().StringP("time-range", "t", "24h", "time range for analysis")

	return cmd
}

// analyzeIdleCmd creates the idle analysis command
func analyzeIdleCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "idle [namespace]",
		Short: "Find idle workloads",
		Long:  "Identify idle workloads using ML-powered analysis",
		RunE: func(cmd *cobra.Command, args []string) error {
			return analyzeIdle(cmd, args)
		},
	}

	// Add flags
	cmd.Flags().Float64P("confidence", "c", 0.85, "confidence threshold")
	cmd.Flags().StringP("time-range", "t", "7d", "time range for analysis")
	cmd.Flags().BoolP("include-health-checks", "h", true, "include health check filtering")

	return cmd
}

// analyzeResourcesCmd creates the resource analysis command
func analyzeResourcesCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "resources [resource-type]",
		Short: "Analyze resource usage",
		Long:  "Analyze CPU, memory, and network usage patterns",
		RunE: func(cmd *cobra.Command, args []string) error {
			return analyzeResources(cmd, args)
		},
	}

	// Add flags
	cmd.Flags().StringP("time-range", "t", "24h", "time range for analysis")
	cmd.Flags().StringP("namespace", "n", "", "namespace to analyze")

	return cmd
}

// analyzeCostCmd creates the cost analysis command
func analyzeCostCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "cost [cluster-name]",
		Short: "Analyze cluster costs",
		Long:  "Analyze cost breakdown and optimization opportunities",
		RunE: func(cmd *cobra.Command, args []string) error {
			return analyzeCost(cmd, args)
		},
	}

	// Add flags
	cmd.Flags().StringP("time-range", "t", "30d", "time range for analysis")
	cmd.Flags().BoolP("detailed", "d", false, "detailed cost breakdown")

	return cmd
}

// analyzePerformanceCmd creates the performance analysis command
func analyzePerformanceCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "performance [cluster-name]",
		Short: "Analyze cluster performance",
		Long:  "Analyze performance metrics and bottlenecks",
		RunE: func(cmd *cobra.Command, args []string) error {
			return analyzePerformance(cmd, args)
		},
	}

	// Add flags
	cmd.Flags().StringP("time-range", "t", "24h", "time range for analysis")
	cmd.Flags().BoolP("detailed", "d", false, "detailed performance analysis")

	return cmd
}

// Implementation functions
func analyzeCluster(cmd *cobra.Command, args []string) error {
	clusterName := "default"
	if len(args) > 0 {
		clusterName = args[0]
	}

	// Get flags
	namespace, _ := cmd.Flags().GetString("namespace")
	timeRange, _ := cmd.Flags().GetString("time-range")
	detailed, _ := cmd.Flags().GetBool("detailed")
	includeCosts, _ := cmd.Flags().GetBool("include-costs")

	// Build arguments
	args = []string{"cluster", clusterName}
	if namespace != "" {
		args = append(args, "--namespace", namespace)
	}
	if timeRange != "" {
		args = append(args, "--time-range", timeRange)
	}
	if detailed {
		args = append(args, "--detailed")
	}
	if includeCosts {
		args = append(args, "--include-costs")
	}

	return executePythonCommand("analyze", args)
}

func analyzePod(cmd *cobra.Command, args []string) error {
	podName := args[0]
	namespace, _ := cmd.Flags().GetString("namespace")
	timeRange, _ := cmd.Flags().GetString("time-range")

	// Build arguments
	cmdArgs := []string{"pod", podName}
	if namespace != "" {
		cmdArgs = append(cmdArgs, "--namespace", namespace)
	}
	if timeRange != "" {
		cmdArgs = append(cmdArgs, "--time-range", timeRange)
	}

	return executePythonCommand("analyze", cmdArgs)
}

func analyzeIdle(cmd *cobra.Command, args []string) error {
	namespace := "default"
	if len(args) > 0 {
		namespace = args[0]
	}

	// Get flags
	confidence, _ := cmd.Flags().GetFloat64("confidence")
	timeRange, _ := cmd.Flags().GetString("time-range")
	includeHealthChecks, _ := cmd.Flags().GetBool("include-health-checks")

	// Build arguments
	cmdArgs := []string{"idle", namespace}
	cmdArgs = append(cmdArgs, "--confidence", fmt.Sprintf("%.2f", confidence))
	if timeRange != "" {
		cmdArgs = append(cmdArgs, "--time-range", timeRange)
	}
	if !includeHealthChecks {
		cmdArgs = append(cmdArgs, "--no-health-check-filtering")
	}

	return executePythonCommand("analyze", cmdArgs)
}

func analyzeResources(cmd *cobra.Command, args []string) error {
	resourceType := "all"
	if len(args) > 0 {
		resourceType = args[0]
	}

	// Get flags
	timeRange, _ := cmd.Flags().GetString("time-range")
	namespace, _ := cmd.Flags().GetString("namespace")

	// Build arguments
	cmdArgs := []string{"resources", resourceType}
	if timeRange != "" {
		cmdArgs = append(cmdArgs, "--time-range", timeRange)
	}
	if namespace != "" {
		cmdArgs = append(cmdArgs, "--namespace", namespace)
	}

	return executePythonCommand("analyze", cmdArgs)
}

func analyzeCost(cmd *cobra.Command, args []string) error {
	clusterName := "default"
	if len(args) > 0 {
		clusterName = args[0]
	}

	// Get flags
	timeRange, _ := cmd.Flags().GetString("time-range")
	detailed, _ := cmd.Flags().GetBool("detailed")

	// Build arguments
	cmdArgs := []string{"cost", clusterName}
	if timeRange != "" {
		cmdArgs = append(cmdArgs, "--time-range", timeRange)
	}
	if detailed {
		cmdArgs = append(cmdArgs, "--detailed")
	}

	return executePythonCommand("analyze", cmdArgs)
}

func analyzePerformance(cmd *cobra.Command, args []string) error {
	clusterName := "default"
	if len(args) > 0 {
		clusterName = args[0]
	}

	// Get flags
	timeRange, _ := cmd.Flags().GetString("time-range")
	detailed, _ := cmd.Flags().GetBool("detailed")

	// Build arguments
	cmdArgs := []string{"performance", clusterName}
	if timeRange != "" {
		cmdArgs = append(cmdArgs, "--time-range", timeRange)
	}
	if detailed {
		cmdArgs = append(cmdArgs, "--detailed")
	}

	return executePythonCommand("analyze", cmdArgs)
}

 

File: api_server/routers/analyze.py
"""
UPID CLI API Server - Analysis Router
Enterprise cluster and workload analysis endpoints
"""

import logging
import asyncio
from datetime import datetime, timedelta
from typing import Dict, Any, List
from fastapi import APIRouter, HTTPException, status, BackgroundTasks
import uuid

from api_server.models.requests import (
    ClusterAnalysisRequest, 
    IdleAnalysisRequest, 
    CostAnalysisRequest
)
from api_server.models.responses import (
    AnalysisResponse, 
    ClusterAnalysisResult, 
    IdleAnalysisResult,
    ResourceMetrics,
    WorkloadInfo,
    IdleWorkload,
    ResponseStatus
)

logger = logging.getLogger(__name__)
router = APIRouter()


# Mock data generators for development (replace with real implementations)
def generate_mock_resource_metrics() -> ResourceMetrics:
    """Generate mock resource metrics"""
    return ResourceMetrics(
        cpu_usage_cores=2.5,
        cpu_usage_percent=62.5,
        memory_usage_bytes=8589934592,  # 8GB
        memory_usage_percent=75.0,
        network_rx_bytes=1073741824,  # 1GB
        network_tx_bytes=536870912,   # 512MB
        storage_usage_bytes=21474836480  # 20GB
    )


def generate_mock_workload_info(name: str, namespace: str = "default") -> WorkloadInfo:
    """Generate mock workload information"""
    return WorkloadInfo(
        name=name,
        namespace=namespace,
        type="Deployment",
        replicas=3,
        status="Running",
        labels={"app": name, "version": "v1.0"},
        annotations={"deployment.kubernetes.io/revision": "1"},
        created_at=datetime.utcnow() - timedelta(days=30),
        metrics=generate_mock_resource_metrics()
    )


@router.post("/cluster", response_model=AnalysisResponse)
async def analyze_cluster(request: ClusterAnalysisRequest, background_tasks: BackgroundTasks):
    """
    Perform comprehensive cluster analysis
    
    Analyzes the entire Kubernetes cluster including:
    - Resource utilization across all nodes and workloads
    - Cost breakdown by namespace and workload type
    - Performance bottlenecks and optimization opportunities
    - Health and efficiency scoring
    """
    try:
        logger.info(f"🔍 Starting cluster analysis for cluster: {request.cluster_id}")
        
        # In production, this would:
        # 1. Connect to Kubernetes cluster using cluster_id
        # 2. Collect real metrics from Prometheus/metrics-server
        # 3. Analyze resource usage patterns
        # 4. Calculate cost implications
        # 5. Generate optimization recommendations
        
        # Mock implementation for demonstration
        mock_workloads = [
            generate_mock_workload_info("web-frontend", "production"),
            generate_mock_workload_info("api-backend", "production"),
            generate_mock_workload_info("database", "production"),
            generate_mock_workload_info("cache-redis", "production"),
            generate_mock_workload_info("monitoring", "system"),
            generate_mock_workload_info("logging", "system")
        ]
        
        mock_cost_analysis = {
            "total_monthly_cost": 2500.0,
            "cost_by_namespace": {
                "production": 2000.0,
                "system": 400.0,
                "default": 100.0
            },
            "cost_by_workload_type": {
                "Deployment": 2200.0,
                "StatefulSet": 200.0,
                "DaemonSet": 100.0
            },
            "cost_trend_30_days": [
                {"date": "2025-01-01", "cost": 2400.0},
                {"date": "2025-01-15", "cost": 2450.0},
                {"date": "2025-01-25", "cost": 2500.0}
            ]
        }
        
        mock_recommendations = [
            {
                "type": "resource_optimization",
                "priority": "high",
                "title": "Optimize CPU requests for web-frontend",
                "description": "CPU requests are over-provisioned by 40%",
                "potential_savings": 200.0,
                "risk_level": "low"
            },
            {
                "type": "idle_workload",
                "priority": "medium", 
                "title": "Consider scaling down cache-redis during off-hours",
                "description": "Cache utilization drops to 5% after business hours",
                "potential_savings": 150.0,
                "risk_level": "medium"
            },
            {
                "type": "zero_pod_scaling",
                "priority": "high",
                "title": "Enable zero-pod scaling for development workloads",
                "description": "Development environment has 60% idle time",
                "potential_savings": 300.0,
                "risk_level": "low"
            }
        ]
        
        analysis_result = ClusterAnalysisResult(
            cluster_id=request.cluster_id,
            analysis_timestamp=datetime.utcnow(),
            cluster_metrics=generate_mock_resource_metrics(),
            workloads=mock_workloads,
            cost_analysis=mock_cost_analysis,
            recommendations=mock_recommendations,
            health_score=85.5,
            efficiency_score=72.3
        )
        
        # Schedule background task for detailed analysis
        background_tasks.add_task(
            perform_detailed_analysis, 
            request.cluster_id, 
            request.time_range_hours
        )
        
        logger.info(f"✅ Cluster analysis completed for: {request.cluster_id}")
        
        return AnalysisResponse(
            status=ResponseStatus.SUCCESS,
            message=f"Cluster analysis completed successfully. Health score: {analysis_result.health_score}%",
            data=analysis_result
        )
        
    except Exception as e:
        logger.error(f"Cluster analysis failed: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Cluster analysis failed: {str(e)}"
        )


@router.post("/idle", response_model=AnalysisResponse)
async def analyze_idle_workloads(request: IdleAnalysisRequest):
    """
    Analyze idle workloads with Health Check Illusion filtering
    
    This is UPID's core differentiator - identifying truly idle workloads
    by filtering out health check traffic that creates false usage signals.
    """
    try:
        logger.info(f"🔍 Starting idle workload analysis for cluster: {request.cluster_id}")
        
        # In production, this would:
        # 1. Connect to cluster and collect metrics
        # 2. Apply Health Check Illusion filtering
        # 3. Use ML models to predict idle probability
        # 4. Calculate potential cost savings
        # 5. Assess optimization risk levels
        
        # Mock idle workloads with realistic data
        mock_idle_workloads = [
            IdleWorkload(
                workload=generate_mock_workload_info("legacy-api", request.namespace),
                idle_confidence=0.92,
                idle_duration_hours=48.5,
                potential_savings_monthly=150.0,
                recommendation="Safe to scale to zero during off-hours (8pm-6am)",
                risk_level="low"
            ),
            IdleWorkload(
                workload=generate_mock_workload_info("batch-processor", request.namespace),
                idle_confidence=0.87,
                idle_duration_hours=72.0,
                potential_savings_monthly=220.0,
                recommendation="Consider zero-pod scaling with traffic-based auto-scaling",
                risk_level="low"
            ),
            IdleWorkload(
                workload=generate_mock_workload_info("dev-environment", request.namespace),
                idle_confidence=0.95,
                idle_duration_hours=120.0,
                potential_savings_monthly=350.0,
                recommendation="Excellent candidate for zero-pod scaling",
                risk_level="very_low"
            )
        ]
        
        # Filter based on confidence threshold
        filtered_workloads = [
            workload for workload in mock_idle_workloads 
            if workload.idle_confidence >= request.confidence_threshold
        ]
        
        total_savings = sum(w.potential_savings_monthly for w in filtered_workloads)
        
        analysis_result = IdleAnalysisResult(
            cluster_id=request.cluster_id,
            namespace=request.namespace,
            analysis_timestamp=datetime.utcnow(),
            time_range_hours=request.time_range_hours,
            total_workloads_analyzed=15,  # Mock total
            idle_workloads=filtered_workloads,
            total_potential_savings=total_savings,
            health_check_traffic_filtered=request.exclude_health_checks
        )
        
        logger.info(
            f"✅ Idle analysis completed: {len(filtered_workloads)} idle workloads found, "
            f"${total_savings:.2f} potential monthly savings"
        )
        
        return AnalysisResponse(
            status=ResponseStatus.SUCCESS,
            message=f"Idle workload analysis completed. Found {len(filtered_workloads)} idle workloads with ${total_savings:.2f} potential monthly savings",
            data=analysis_result
        )
        
    except Exception as e:
        logger.error(f"Idle analysis failed: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Idle workload analysis failed: {str(e)}"
        )


@router.post("/cost", response_model=AnalysisResponse)
async def analyze_costs(request: CostAnalysisRequest):
    """
    Perform detailed cost analysis and projections
    
    Analyzes current and projected costs across multiple dimensions:
    - Historical cost trends
    - Cost breakdown by namespace, workload, and resource type
    - Future cost projections based on usage patterns
    - ROI analysis for optimization opportunities
    """
    try:
        logger.info(f"💰 Starting cost analysis for cluster: {request.cluster_id}")
        
        # Mock cost analysis data
        cost_data = {
            "cluster_id": request.cluster_id,
            "analysis_period_days": request.time_range_days,
            "total_current_cost": 2500.0,
            "projected_monthly_cost": 2650.0,
            "cost_trend": "increasing",
            "cost_breakdown": {
                "compute": 1800.0,
                "storage": 400.0,
                "network": 200.0,
                "other": 100.0
            },
            "top_cost_drivers": [
                {"resource": "production/web-frontend", "cost": 500.0, "percentage": 20.0},
                {"resource": "production/database", "cost": 400.0, "percentage": 16.0},
                {"resource": "production/api-backend", "cost": 350.0, "percentage": 14.0}
            ],
            "optimization_opportunities": [
                {
                    "type": "right_sizing",
                    "potential_savings": 400.0,
                    "confidence": 0.85,
                    "effort": "low"
                },
                {
                    "type": "idle_removal",
                    "potential_savings": 300.0,
                    "confidence": 0.92,
                    "effort": "medium"
                }
            ],
            "projections": {
                "1_month": 2650.0,
                "3_months": 7950.0,
                "6_months": 15900.0,
                "12_months": 31800.0
            } if request.include_projections else None
        }
        
        logger.info(f"✅ Cost analysis completed: ${cost_data['total_current_cost']:.2f} current monthly cost")
        
        return AnalysisResponse(
            status=ResponseStatus.SUCCESS,
            message=f"Cost analysis completed. Current monthly cost: ${cost_data['total_current_cost']:.2f}",
            data=cost_data
        )
        
    except Exception as e:
        logger.error(f"Cost analysis failed: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Cost analysis failed: {str(e)}"
        )


@router.get("/history/{cluster_id}", response_model=AnalysisResponse)
async def get_analysis_history(cluster_id: str, limit: int = 10):
    """
    Get historical analysis results for a cluster
    
    Returns previous analysis results to track trends and improvements over time.
    """
    try:
        logger.info(f"📊 Retrieving analysis history for cluster: {cluster_id}")
        
        # Mock historical data
        history_data = {
            "cluster_id": cluster_id,
            "analysis_history": [
                {
                    "timestamp": "2025-01-25T10:00:00Z",
                    "type": "cluster_analysis",
                    "health_score": 85.5,
                    "efficiency_score": 72.3,
                    "cost": 2500.0
                },
                {
                    "timestamp": "2025-01-24T10:00:00Z", 
                    "type": "idle_analysis",
                    "idle_workloads_found": 3,
                    "potential_savings": 720.0
                },
                {
                    "timestamp": "2025-01-23T10:00:00Z",
                    "type": "cluster_analysis", 
                    "health_score": 83.2,
                    "efficiency_score": 70.1,
                    "cost": 2480.0
                }
            ][:limit],
            "trends": {
                "health_score_trend": "improving",
                "cost_trend": "stable",
                "efficiency_trend": "improving"
            }
        }
        
        return AnalysisResponse(
            status=ResponseStatus.SUCCESS,
            message=f"Retrieved {len(history_data['analysis_history'])} historical analysis records",
            data=history_data
        )
        
    except Exception as e:
        logger.error(f"Failed to retrieve analysis history: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to retrieve analysis history: {str(e)}"
        )


async def perform_detailed_analysis(cluster_id: str, time_range_hours: int):
    """
    Background task for detailed analysis
    
    This runs comprehensive analysis that may take longer,
    including ML model predictions and deep resource analysis.
    """
    try:
        logger.info(f"🔄 Starting detailed background analysis for cluster: {cluster_id}")
        
        # Simulate long-running analysis
        await asyncio.sleep(5)
        
        # In production, this would:
        # 1. Run ML models for predictions
        # 2. Perform deep resource analysis  
        # 3. Generate detailed optimization recommendations
        # 4. Store results in database for later retrieval
        
        logger.info(f"✅ Detailed background analysis completed for cluster: {cluster_id}")
        
    except Exception as e:
        logger.error(f"Background analysis failed for cluster {cluster_id}: {e}")